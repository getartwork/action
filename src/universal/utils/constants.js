/**
 * Big stuff:
 */
export const APP_CDN_USER_ASSET_SUBDIR = '/store';
export const APP_MAX_AVATAR_FILE_SIZE = 1024 * 1024;
export const APP_NAME = 'Action';
export const APP_REDUX_KEY = `${APP_NAME}:redux`;

/**
 * Upgrade pending states, called when the server version changes.
 * Must be a string:
 *
 *    APP_UPGRADE_PENDING_FALSE: no upgrade needed
 *   APP_UPGRADE_PENDING_RELOAD: client reload needed
 *     APP_UPGRADE_PENDING_DONE: upgrade complete
 *
 * Stored in localSession by APP_UPGRADE_PENDING_KEY.
 */
export const APP_UPGRADE_PENDING_KEY = `${APP_NAME}:upgradePending`;
export const APP_UPGRADE_PENDING_FALSE = 'false';
export const APP_UPGRADE_PENDING_RELOAD = 'reload';
export const APP_UPGRADE_PENDING_DONE = 'done';
export const APP_VERSION_KEY = `${APP_NAME}:version`;         // in localStorage
export const APP_WEBPACK_PUBLIC_PATH_DEFAULT = '/static/';

/* Phases */
export const LOBBY = 'lobby';
// lowercase here to match url
export const CHECKIN = 'checkin';
export const UPDATES = 'updates';
export const FIRST_CALL = 'firstcall';
export const AGENDA_ITEMS = 'agendaitems';
export const LAST_CALL = 'lastcall';
export const SUMMARY = 'summary';
export const phaseArray = [LOBBY, CHECKIN, UPDATES, FIRST_CALL, AGENDA_ITEMS, LAST_CALL, SUMMARY];

/* Columns */
export const ACTIVE = 'active';
export const STUCK = 'stuck';
export const DONE = 'done';
export const FUTURE = 'future';
export const columnArray = [FUTURE, STUCK, ACTIVE, DONE];
export const meetingColumnArray = [DONE, ACTIVE, STUCK, FUTURE];

/* Drag-n-Drop Items*/
export const ACTION = 'action';
export const PROJECT = 'project';
export const AGENDA_ITEM = 'agendaItem';

/* Sorting */
export const SORT_STEP = 1;
export const DND_THROTTLE = 25;

/* Areas */
export const MEETING = 'meeting';
export const TEAM_DASH = 'teamDash';
export const USER_DASH = 'userDash';

/* NotificationRow Types */
// Sent to Billing Leaders when their trial will expire in < 2 weeks
// varList = [trialExpiresAt]
export const TRIAL_EXPIRES_SOON = 'TRIAL_EXPIRES_SOON';
// Sent to Billing Leaders when their trial has expired
// varList = [trialExpiresAt]
export const TRIAL_EXPIRED = 'TRIAL_EXPIRED';
// Sent to Billing Leaders when a reoccuring payment gets rejected
// varList = [last4, brand]
export const PAYMENT_REJECTED = 'PAYMENT_REJECTED';
// Sent to Billing Leaders when an orgMember attempts to add a non-orgMember to a team
// [inviterId, inviterName, inviteeEmail, invitedTeamId, invitedTeamName]
export const REQUEST_NEW_USER = 'REQUEST_NEW_USER';
// Sent to the orgMember that generated the new user request
// [reason, billingLeaderName, inviteeEmail]
export const DENY_NEW_USER = 'DENY_NEW_USER';
// sent to the orgMember that just got promoted, goes away if they get demoted before acknowledging it
// [orgName]
export const PROMOTE_TO_BILLING_LEADER = 'PROMOTE_TO_BILLING_LEADER';
// sent to members of team that was archived
export const TEAM_ARCHIVED = 'TEAM_ARCHIVED';

export const notificationTypes = [
  TRIAL_EXPIRES_SOON,
  TRIAL_EXPIRED,
  PAYMENT_REJECTED,
  PROMOTE_TO_BILLING_LEADER,
  REQUEST_NEW_USER,
  DENY_NEW_USER,
  TEAM_ARCHIVED
];

export const billingLeaderTypes = [
  TRIAL_EXPIRES_SOON,
  TRIAL_EXPIRED,
  PAYMENT_REJECTED,
  REQUEST_NEW_USER
];

/* User Settings */
export const SETTINGS = 'settings';
export const ORGANIZATIONS = 'organizations';
export const NOTIFICATIONS = 'notifications';
export const settingsOrder = [SETTINGS, ORGANIZATIONS, NOTIFICATIONS];

/* Org Settings */
export const BILLING_PAGE = 'billing';
export const MEMBERS_PAGE = 'members';

/* User Org Roles */
export const BILLING_LEADER = 'billingLeader';

/* Stripe */
// changing this does NOT change it in stripe, it just changes the UI
export const MONTHLY_PRICE = 5;
export const ADDED_USERS = 'ADDED_USERS';
export const REMOVED_USERS = 'REMOVED_USERS';
export const INACTIVITY_ADJUSTMENTS = 'INACTIVITY_ADJUSTMENTS';
export const OTHER_ADJUSTMENTS = 'OTHER_ADJUSTMENTS';

/* Invoice status variables */
export const UPCOMING = 'UPCOMING';
export const PENDING = 'PENDING';
export const PAID = 'PAID';
export const FAILED = 'FAILED';

/* character limits */
export const PROJECT_MAX_CHARS = 512;
