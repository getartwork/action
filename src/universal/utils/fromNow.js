export const thresholds = {
  second: 1000,
  minute: 60000,
  hour: 3600000,
  day: 86400000,
  week: 604800000,
  month: 2592000000,
  year: 31536000000,
  inf: Infinity
};

export default function fromNow(maybeTime) {
  const time = new Date(maybeTime);
  const now = Date.now();
  const distance = Math.abs(now - time) || 0;
  const ago = now > time ? ' ago' : '';
  if (distance < 1000) return 'just now';
  const threshKeys = Object.keys(thresholds);
  let prevThresh = 1000;
  for (let i = 1; i < threshKeys.length; i++) {
    const thresh = thresholds[threshKeys[i]];
    if (distance < thresh) {
      const roundDistance = Math.round(distance / prevThresh);
      const units = `${threshKeys[i - 1]}${roundDistance === 1 ? '' : 's'}${ago}`;
      return `${roundDistance} ${units}`;
    }
    prevThresh = thresh;
  }
  throw new Error('Infinite timestamp calculated!');
}
