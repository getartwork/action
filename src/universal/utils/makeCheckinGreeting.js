const greetings = [
  'Hola',
  'Bonjour',
  'Guten tag',
  'Ciao',
  'Namaste',
  'Jambo',
  'Oi',
  'Salut',
  'Hallo',
  'Hej',
  'Nǐ hǎo',
  'Yeoboseyo',
  'Shalom',
  'Merhaba',
  'Zdravo'
];

const questions = [
  'What’s your spirit animal today',
  'What’s something you’re looking forward to',
  'What’s something you’re worried about',
  'What’s something you’ve learned recently',
  'Anything inspire you recently',
  'What’s the weirdest thing in your fridge right now',
  'How are your hobbies doing',
  'What are you cooking next',
  'Learn any new slang lately',
  'Heard any good advice lately',
  'If you could have one superpower today, what would it be'
];

export const makeCheckinGreeting = (week) => {
  return greetings[week % greetings.length];
};

export const makeCheckinQuestion = (week) => {
  return questions[week % questions.length];
};
