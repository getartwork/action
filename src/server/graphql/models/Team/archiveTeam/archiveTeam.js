import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import getRethink from 'server/database/rethinkDriver';
import {
  requireSUOrLead,
  requireWebsocket
} from 'server/utils/authorization';
import {
  GraphQLNonNull,
  GraphQLBoolean,
  GraphQLString
} from 'graphql';
import shortid from 'shortid';
import {KICK_OUT, USER_MEMO} from 'universal/subscriptions/constants';
import {TEAM_ARCHIVED} from 'universal/utils/constants';

export default {
  type: GraphQLBoolean,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'The teamId to archive (or delete, if team is unused)'
    }
  },
  async resolve(source, {teamId}, {authToken, exchange, socket}) {
    const r = getRethink();
    const now = new Date();

    // AUTH
    requireSUOrLead(authToken, teamId);
    requireWebsocket(socket);

    // RESOLUTION
    const dbResult = await r.table('Team')
      .get(teamId)
      .pluck('name', 'orgId')
      .do((team) => ({
        projectCount: r.table('Project').getAll(teamId, {index: 'teamId'}).count(),
        team,
        userIds: r.table('TeamMember').getAll(teamId, {index: 'teamId'})('userId').coerceTo('array')
      }))
      .do((doc) => r.table('User')
        .getAll(r.args(doc('userIds')), {index: 'id'})
        .update((user) => ({tms: user('tms').difference([teamId])}), {returnChanges: true})
        .do((userResult) => r.branch(
          r.and(doc('projectCount').eq(0), doc('userIds').count().eq(1)),
          {
            // Team has no projects nor addn'l TeamMembers, hard delete it:
            team: doc('team'),
            teamResult: r.table('Team').get(teamId).delete(),
            teamMemberResult: r.table('TeamMember').getAll(teamId, {index: 'teamId'}).delete(),
            userResult,
          },
          {
            // Team has data or TeamMembers, archive team:
            notificationResult: r.table('Notification').insert({
              id: shortid.generate(),
              orgId: doc('team')('orgId'),
              startAt: now,
              type: TEAM_ARCHIVED,
              userIds: doc('userIds'),
              varList: [doc('team')('name')]
            }),
            team: doc('team'),
            teamResult: r.table('Team').get(teamId).update({isArchived: true}),
            userResult
          }
        ))
      );

    const {team: {name: teamName}, userResult: {changes: userChanges}} = dbResult;
    userChanges.forEach((change) => {
      const {id, tms} = change.new_val;
      // update the tms on auth0 in async
      auth0ManagementClient.users.updateAppMetadata({id}, {tms});
      // update the server socket, if they're logged in
      const channel = `${USER_MEMO}/${id}`;
      exchange.publish(channel, {type: KICK_OUT, teamId, teamName});
    });

    return true;
  }
};
