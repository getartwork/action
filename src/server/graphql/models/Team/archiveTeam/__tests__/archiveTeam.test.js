import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import getRethink from 'server/database/rethinkDriver';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import MockDB from 'server/__tests__/setup/MockDB';
import archiveTeam from 'server/graphql/models/Team/archiveTeam/archiveTeam';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import exchange from 'server/__mocks__/exchange';
import socket from 'server/__mocks__/socket';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';

MockDate.set(__now);
console.error = jest.fn();

describe('ArchiveTeam', () => {
  test('archives a team', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {user, team: [updatedTeam], teamMember} = await mockDB.init();
    updatedTeam.isArchived = true;
    const teamLeadId = teamMember.find((tm) => tm.teamId === updatedTeam.id && tm.isLead).userId;
    const teamLead = user.find((usr) => usr.id === teamLeadId);
    const authToken = mockAuthToken(teamLead);
    auth0ManagementClient.__initMock(mockDB.db);
    auth0ManagementClient.users.updateAppMetadata.mockReset();
    exchange.publish.mockReset();

    // TEST
    await archiveTeam.resolve(undefined, {teamId: updatedTeam.id}, {authToken, exchange, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(updatedTeam.orgId, {index: 'orgId'}).orderBy('startAt'),
      team: r.table('Team').get(updatedTeam.id)
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(auth0ManagementClient.users.updateAppMetadata.mock.calls.length)
      .toBe(teamMember.length);
    expect(exchange.publish.mock.calls.length).toBe(teamMember.length);
  });

  test('deletes a team when it has no projects or other team members', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {user: [user], team: [updatedTeam]} = await mockDB
      .newOrg({name: 'Sad Sacks, Inc.'})
      .newTeam({name: 'The Lonely Ones'})
      .newUser({name: 'Leader of One'})
      .newTeamMember({isLead: true});
    updatedTeam.isArchived = true;
    const authToken = mockAuthToken(user);
    auth0ManagementClient.__initMock(mockDB.db);
    auth0ManagementClient.users.updateAppMetadata.mockReset();
    exchange.publish.mockReset();

    // TEST
    await archiveTeam.resolve(undefined, {teamId: updatedTeam.id}, {authToken, exchange, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(updatedTeam.orgId, {index: 'orgId'}).orderBy('startAt'),
      team: r.table('Team').get(updatedTeam.id).default({}),
      teamMember: r.table('TeamMember').getAll(updatedTeam.id, {index: 'teamId'}).default([]),
      user: r.table('User').get(user.id)
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(auth0ManagementClient.users.updateAppMetadata.mock.calls.length)
      .toBe(1);
    expect(exchange.publish.mock.calls.length).toBe(1);
  });
});
