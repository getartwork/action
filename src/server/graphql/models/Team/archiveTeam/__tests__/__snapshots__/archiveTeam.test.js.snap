// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchiveTeam archives a team 1`] = `
Object {
  "notification": Array [
    Object {
      "id": 0,
      "orgId": 1,
      "startAt": 2017-01-01T08:00:00.000Z,
      "type": "TEAM_ARCHIVED",
      "userIds": Array [
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
      ],
      "varList": Array [
        "Team America",
      ],
    },
  ],
  "team": Object {
    "activeFacilitator": null,
    "facilitatorPhase": "lobby",
    "facilitatorPhaseItem": null,
    "id": 10,
    "isArchived": true,
    "isPaid": true,
    "meetingId": null,
    "meetingPhase": "lobby",
    "meetingPhaseItem": null,
    "name": "Team America",
    "orgId": 1,
  },
}
`;

exports[`ArchiveTeam deletes a team when it has no projects or other team members 1`] = `
Object {
  "notification": Array [],
  "team": Object {},
  "teamMember": Array [],
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "Leader of One",
    "picture": "https://placeimg.com/100/100/animals",
    "tms": Array [],
    "updatedAt": 2017-01-01T07:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 1,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;
